src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1766 @@
+        }
+
+        void setEra(String & text) // NOLINT
+        {
+            boost::to_lower(text);
+            if (text == "bc")
+                throw Exception(ErrorCodes::CANNOT_PARSE_DATETIME, "Era BC exceeds the range of DateTime");
+            else if (text != "ad")
+                throw Exception(ErrorCodes::CANNOT_PARSE_DATETIME, "Unknown era {}", text);
+        }
+
+        ALWAYS_INLINE void setAMPM(String & text)
+        {
+            boost::to_lower(text);
+            if (text == "am")
+                is_am = true;
+            else if (text == "pm")
+                is_am = false;
+            else
+                throw Exception(ErrorCodes::CANNOT_PARSE_DATETIME, "Unknown half day of day: {}", text);
+        }
+
+        ALWAYS_INLINE void setHour(Int32 hour_, bool is_hour_of_half_day_ = false, bool is_clock_hour_ = false)
+        {
+            Int32 max_hour;
+            Int32 min_hour;
+            Int32 new_hour = hour_;
+            if (!is_hour_of_half_day_ && !is_clock_hour_)
+            {
+                max_hour = 23;
+                min_hour = 0;
+            }
+            else if (!is_hour_of_half_day_ && is_clock_hour_)
+            {
+                max_hour = 24;
+                min_hour = 1;
+                new_hour = hour_ % 24;
+            }
+            else if (is_hour_of_half_day_ && !is_clock_hour_)
+            {
+                max_hour = 11;
+                min_hour = 0;
+            }
+            else
+            {
+                max_hour = 12;
+                min_hour = 1;
+                new_hour = hour_ % 12;
+            }
+
+            if (hour_ < min_hour || hour_ > max_hour)
+                throw Exception(
+                    ErrorCodes::CANNOT_PARSE_DATETIME,
+                    "Value {} for hour must be in the range [{}, {}] if_hour_of_half_day={} and is_clock_hour={}",
+                    hour,
+                    max_hour,
+                    min_hour,
+                    is_hour_of_half_day_,
+                    is_clock_hour_);
+
+            hour = new_hour;
+            is_hour_of_half_day = is_hour_of_half_day_;
+            is_clock_hour = is_clock_hour_;
+        }
+
+        /// For debug
+        [[maybe_unused]] String toString() const
+        {
+            String res;
+            res += "year:" + std::to_string(year);
+            res += ",";
+            res += "month:" + std::to_string(month);
+            res += ",";
+            res += "day:" + std::to_string(day);
+            res += ",";
+            res += "hour:" + std::to_string(hour);
+            res += ",";
+            res += "minute:" + std::to_string(minute);
+            res += ",";
+            res += "second:" + std::to_string(second);
+            res += ",";
+            res += "AM:" + std::to_string(is_am);
+            return res;
+        }
+
+        static bool isLeapYear(Int32 year_) { return year_ % 4 == 0 && (year_ % 100 != 0 || year_ % 400 == 0); }
```

Comment:
Would you like to double-check this formula? A year is a leap year if can be divided by 4.
EXCEPTION 1: If it can be divided by 4 and by 100 it is NEVER a leap year.
EXCEPTION 2: If it can be divided by 400 it is ALWAYS a leap year.

--> `(year_ % 4 == 0 && year_ % 100 != 0) || (year_ % 400 == 0)`

Reply:
You can validate it in the test using SQL )

```sql
select year(today()) YEAR,
    if ( (YEAR % 4 = 0 AND YEAR % 100 <> 0) OR YEAR % 400 = 0, 'LEAP', 'Not LEAP' ) leap;
```